ui/radar/radar.ts

```diff
@@ -0,0 +1,499 @@
+        'rank': hunt.rank,
+        'hp': parseFloat(matches.hp),
+        'currentHp': parseFloat(matches.hp),
+        'battleTime': 0,
+        'pos': new Point2D(parseFloat(matches.x), parseFloat(matches.y)),
+        'posZ': parseFloat(matches.z),
+        'addTime': +Date.now(),
+        'dom': tr,
+        'puller': undefined,
+        'skipPuller': matches.hp !== matches.currentHp, // already pulled before being detected
+      };
+      this.targetMonsters[mobKey] = m;
+      this.UpdateMonsterDom(m);
+
+      const mapX = posToMap(m.pos.x).toFixed(1);
+      const mapY = posToMap(m.pos.y).toFixed(1);
+      console.log(`Found: ${m.name} (${mapX}, ${mapY})`);
+
+      PlaySound(m, options);
+    }
+  }
+
+  UpdateMonsterPuller(monster: Monster, puller: string) {
+    if (!this.options.Puller)
+      return;
+    if (monster.puller || monster.skipPuller)
+      return;
+    monster.puller = puller;
+    this.UpdateMonsterDom(monster);
+    console.log(`Pulled: ${puller} => ${monster.name}`);
+  }
+
+  UpdateMonsterDom(monster: Monster) {
+    // Wait for OnPlayerChange.
+    if (!this.playerPos)
+      return;
+
+    let options = this.options;
+    if (monster.rank && monster.rank in options.RankOptions)
+      options = Object.assign({}, this.options, options.RankOptions[monster.rank]);
+    const tr = monster.dom;
+    // calculate rotation based on facing
+    const playerVector = new Point2D(this.playerPos.x, this.playerPos.y);
+    const targetVector = monster.pos;
+    const deltaVector = new Point2D(
+        targetVector.x - playerVector.x,
+        targetVector.y - playerVector.y,
+    );
+    if (tr) {
+      const node = tr.childNodes[1];
+      if (node && node instanceof HTMLElement) {
+        node.innerHTML = `${monster.rank ?? ''}&nbsp;&nbsp;&nbsp;&nbsp;${monster.name}`;
+        if (Math.abs(this.playerPos.z - monster.posZ) > 5)
+          node.innerHTML += '&nbsp;&nbsp;' + (this.playerPos.z < monster.posZ ? '↑' : '↓');
+        node.innerHTML += '<br>' + deltaVector.length().toFixed(2) + 'm';
+        if (Date.now() / 1000 <= monster.battleTime + 60) {
+          node.innerHTML += ' ' + (monster.currentHp * 100 /
+            monster.hp).toFixed(2) + '%';
+        }
+        if (monster.puller)
+          node.innerHTML += '&nbsp;&nbsp;' + monster.puller;
+        // Z position is relative to the map so it's omitted.
+        if (options.Position) {
+          node.innerHTML += '<br>X: ' +
+            posToMap(monster.pos.x).toFixed(1) + '&nbsp;&nbsp;Y:' +
+            posToMap(monster.pos.y).toFixed(1);
+        }
+      }
+    }
+    if (options.DetectionRange > 0 && deltaVector.length() > options.DetectionRange)
+      monster.dom.classList.add('hide');
+    else
+      monster.dom.classList.remove('hide');
+
+    let deltaTheta = Math.atan2(deltaVector.y, deltaVector.x);
+    deltaTheta -= Math.PI - this.playerRotation;
+    const angle = deltaTheta * 180 / Math.PI;
+    const arrowId = 'arrow-' + monster.id;
```

Comment:
```suggestion
    const arrowId = `arrow-${monster.id}`;
```
Surprised this didn't complain?